File: cpu/rust/src/binary_search.rs
//! Binary Search Benchmark — Rust
//! ================================
//! Benchmarks `slice::binary_search` on sorted arrays.
//! Tests memory access patterns and branch prediction.

use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

/// Binary search returning whether the key was found.
pub fn binary_search_sorted(data: &[i32], queries: &[i32]) -> i64 {
    let mut found: i64 = 0;
    for &q in queries {
        if data.binary_search(&q).is_ok() {
            found += 1;
        }
    }
    found
}

/// Lower-bound search returning sum of insertion indices.
pub fn lower_bound_sorted(data: &[i32], queries: &[i32]) -> i64 {
    let mut sum: i64 = 0;
    for &q in queries {
        let idx = data.partition_point(|x| *x < q);
        sum += idx as i64;
    }
    sum
}

/// Generate sorted array and random queries.
pub fn generate_search_data(n: usize, num_queries: usize) -> (Vec<i32>, Vec<i32>) {
    let data: Vec<i32> = (0..n as i32).collect();
    let mut rng = StdRng::seed_from_u64(42);
    let max_val = (n as f64 * 1.2) as i32;
    let queries: Vec<i32> = (0..num_queries).map(|_| rng.gen_range(0..max_val)).collect();
    (data, queries)
}

File: cpu/cpp/src/binary_search.cpp
/**
 * Binary Search Benchmark — C++
 * ===============================
 * Benchmarks std::binary_search and std::lower_bound on sorted arrays.
 * Tests memory access patterns and branch prediction.
 * Input sizes: 1M, 10M, 100M sorted elements with 100K random lookups each.
 */

#include <algorithm>
#include <benchmark/benchmark.h>
#include <cstdint>
#include <numeric>
#include <random>
#include <vector>

#include "benchmark_utils.h"

// ── Benchmark: std::binary_search ───────────────────────────────────────────

static void BM_BinarySearch(benchmark::State& state) {
    const size_t n = state.range(0);
    const size_t num_queries = 100'000;

    // Create sorted array
    std::vector<int32_t> data(n);
    std::iota(data.begin(), data.end(), 0);

    // Generate random search targets (mix of hits and misses)
    std::mt19937 rng(42);
    std::uniform_int_distribution<int32_t> dist(0, static_cast<int32_t>(n * 1.2));
    std::vector<int32_t> queries(num_queries);
    for (auto& q : queries) q = dist(rng);

    for (auto _ : state) {
        int64_t found = 0;
        for (const auto& q : queries) {
            found += std::binary_search(data.begin(), data.end(), q);
        }
        benchmark::DoNotOptimize(found);
    }

    state.SetItemsProcessed(static_cast<int64_t>(state.iterations()) * num_queries);
    state.SetLabel("std::binary_search");
}

// ── Benchmark: std::lower_bound ─────────────────────────────────────────────

static void BM_LowerBound(benchmark::State& state) {
    const size_t n = state.range(0);
    const size_t num_queries = 100'000;

    std::vector<int32_t> data(n);
    std::iota(data.begin(), data.end(), 0);

    std::mt19937 rng(42);
    std::uniform_int_distribution<int32_t> dist(0, static_cast<int32_t>(n * 1.2));
    std::vector<int32_t> queries(num_queries);
    for (auto& q : queries) q = dist(rng);

    for (auto _ : state) {
        int64_t sum_indices = 0;
        for (const auto& q : queries) {
            auto it = std::lower_bound(data.begin(), data.end(), q);
            sum_indices += std::distance(data.begin(), it);
        }
        benchmark::DoNotOptimize(sum_indices);
    }

    state.SetItemsProcessed(static_cast<int64_t>(state.iterations()) * num_queries);
    state.SetLabel("std::lower_bound");
}

// ── Register ────────────────────────────────────────────────────────────────

BENCHMARK(BM_BinarySearch)
    ->Unit(benchmark::kMillisecond)
    ->Arg(1'000'000)
    ->Arg(10'000'000)
    ->Arg(100'000'000)
    ->MinTime(bench_utils::cpu_min_time_seconds())
    ->Repetitions(bench_utils::cpu_repetitions(30))
    ->ReportAggregatesOnly(true)
    ->DisplayAggregatesOnly(true);

BENCHMARK(BM_LowerBound)
    ->Unit(benchmark::kMillisecond)
    ->Arg(1'000'000)
    ->Arg(10'000'000)
    ->Arg(100'000'000)
    ->MinTime(bench_utils::cpu_min_time_seconds())
    ->Repetitions(bench_utils::cpu_repetitions(30))
    ->ReportAggregatesOnly(true)
    ->DisplayAggregatesOnly(true);
